### PR TITLE
[G-API]: Performance tests for kmeans

### PR DESCRIPTION
**Pull request by OrestChura**
- Performance tests for kmeans operation added
 - New file created to share common parts between accuracy and performance tests for core kernels
 - Accuracy test bodies refactored
 - Some test typos corrected

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

build_image:Custom=centos:7
buildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

xbuild_image:Custom=ubuntu-openvino-2021.1.0:20.04
xbuild_image:Custom Win=openvino-2021.1.0
xbuild_image:Custom Mac=openvino-2021.1.0

test_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by OrestChura** [comment]
@aDanPin please take a look

**Review by aDanPin — APPROVED** [review]
Everything except for two small comments looks good to me
Have a good merge!

File: modules/gapi/test/common/gapi_core_tests_common.hpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+template<typename Labels, typename In>
+void kmeansTestOpenCVCompare(const In& in, const Labels& bestLabels, const int K,
+                             const cv::KmeansFlags flags, const double compact_gapi,
+                             const Labels& labels_gapi, const In& centers_gapi,
+                             const CompareMats& cmpF = AbsExact().to_compare_obj())
+{
+    const cv::TermCriteria criteria(cv::TermCriteria::MAX_ITER + cv::TermCriteria::EPS, 30, 0);
+    const int attempts = 1;
+    double compact_ocv = -1.;
```
Comment (mpashchenkov) [inline modules/gapi/test/common/gapi_core_tests_common.hpp:129]: You can drop this step and create this variable as result of cv::kmeans.

File: modules/gapi/test/common/gapi_core_tests_common.hpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    double compact_gapi = -1.;
+    Labels labels_gapi;
+    In centers_gapi;
+    if (flags & cv::KMEANS_USE_INITIAL_LABELS)
+    {
+        Labels bestLabels;
+        { // step to generalize cv::Mat & std::vector cases of bestLabels' types
+            const int amount = (sz.height != 1 || sz.width == -1) ? sz.height : sz.width;
+            cv::Mat bestLabelsMat(cv::Size{1, amount}, CV_32SC1);
+            cv::randu(bestLabelsMat, 0, K);
```
Comment (mpashchenkov) [inline modules/gapi/test/common/gapi_core_tests_common.hpp:163]: Should it be a test parameter instead randu?

File: modules/gapi/test/common/gapi_tests_common.hpp [hunk 3]
```diff
@@ -334,7 +334,7 @@ class TestFunctional
 #endif
     // initialize std::vector<cv::Point_<T>>/std::vector<cv::Point3_<T>>
     template <typename T, template <typename> class Pt>
-    void initPointsVectorRandU(const int sz_in, std::vector<Pt<T>> &vec_)
+    void initPointsVectorRandU(const int sz_in, std::vector<Pt<T>> &vec_) const
```
Comment (mpashchenkov) [inline modules/gapi/test/common/gapi_tests_common.hpp:337]: Why did you add the const?

File: modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp [hunk 4]
```diff
@@ -282,6 +282,53 @@ INSTANTIATE_TEST_CASE_P(ConvertToPerfTestCPU, ConvertToPerfTest,
             Values(0.0),
             Values(cv::compile_args(CORE_CPU))));
 
+INSTANTIATE_TEST_CASE_P(KMeansNDNoInitPerfTestCPU, KMeansNDPerfTest,
+                        Combine(Values(CV_32FC1, CV_32FC3),
+                                Values(cv::Size(1, 20),
+                                       cv::Size(2, 20),
```
Comment (mpashchenkov) [inline modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp:288]: I think, you can avoid an intermediate value for perf tests.

File: modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp [hunk 5]
```diff
@@ -282,6 +282,53 @@ INSTANTIATE_TEST_CASE_P(ConvertToPerfTestCPU, ConvertToPerfTest,
...
+INSTANTIATE_TEST_CASE_P(KMeans3DNoInitPerfTestCPU, KMeans3DPerfTest,
+                        Combine(Values(20, 720),
+                                Values(5, 15),
+                                Values(cv::KMEANS_RANDOM_CENTERS, cv::KMEANS_PP_CENTERS),
+                                Values(cv::compile_args(CORE_CPU))));
+
+INSTANTIATE_TEST_CASE_P(KMeans3DInitPerfTestCPU, KMeans3DPerfTest,
+                        Combine(Values(20, 720),
+                                Values(5, 15),
+                                Values(cv::KMEANS_RANDOM_CENTERS | cv::KMEANS_USE_INITIAL_LABELS,
```
Comment (mpashchenkov) [inline modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp:328]: One more time, do these values depend?

**Review by mpashchenkov — COMMENTED** [review]
First wave. Thanks.

File: modules/gapi/perf/common/gapi_core_perf_tests_inl.hpp [hunk 6]
```diff
@@ -1905,6 +1907,135 @@ PERF_TEST_P_(ConvertToPerfTest, TestPerformance)
 
 //------------------------------------------------------------------------------
 
+PERF_TEST_P_(KMeansNDPerfTest, TestPerformance)
+{
+    MatType2 type = -1;
+    cv::Size sz;
+    CompareMats cmpF;
+    int K = -1;
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/perf/common/gapi_core_perf_tests_inl.hpp:1915]: Yes, it is. The naming of variables is used after the OpenCV's `cv::kmeans` function which in particular has the `K` parameter

File: modules/gapi/test/common/gapi_core_tests_common.hpp (hunk 2 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_core_tests_common.hpp:163]: Sorry, what do you mean? `bestLabelsMat` is the input `cv::Mat`, so I fill it randomly

File: modules/gapi/test/common/gapi_tests_common.hpp (hunk 3 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_tests_common.hpp:337]: It could be needed in the previous versions of the code, if I wanted to pass the const reference of the test object to a function and initialize a vector inside it.

Now it isn't actually used, but I decided to preserve this change as in fact this method is `const` because it doesn't change the class's fields. So, I made a mistake not to add the `const` when I wrote this method in the past and want to correct it here.

File: modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp (hunk 5 above)
Comment (OrestChura) [reply modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp:328]: Yes, `cv::KMEANS_USE_INITIAL_LABELS` flags should be tested as they can influence the performance. As it is written in the docs, they allow, `during the first (and possibly the only) attempt, use the user-supplied labels instead of computing them from the initial centers.`

File: modules/gapi/test/common/gapi_core_tests_common.hpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    // OpenCV code /////////////////////////////////////////////////////////////
+    compact_ocv = cv::kmeans(in, K, labels_ocv, criteria, attempts, flags, centers_ocv);
+    // Comparison //////////////////////////////////////////////////////////////
+    EXPECT_TRUE(compact_gapi == compact_ocv);
+    EXPECT_TRUE(compareKMeansOutputs(labels_gapi, labels_ocv, cmpF));
+    EXPECT_TRUE(compareKMeansOutputs(centers_gapi, centers_ocv, cmpF));
+}
+
+template<typename In>
+using KMeansLabelType = typename std::conditional<std::is_same<In, cv::Mat>::value,
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/common/gapi_core_tests_common.hpp:146]: Thanks!
OK, comment added

File: modules/gapi/perf/common/gapi_core_perf_tests_inl.hpp [hunk 8]
```diff
@@ -1905,6 +1907,135 @@ PERF_TEST_P_(ConvertToPerfTest, TestPerformance)
...
+    cv::GCompileArgs compile_args;
+    std::tie(type, sz, cmpF, K, flags, compile_args) = GetParam();
+
+    initMatrixRandU(type, sz, -1, false);
+
+    double compact_gapi = -1.;
+    cv::Mat labels_gapi, centers_gapi;
+    if (flags & cv::KMEANS_USE_INITIAL_LABELS)
+    {
+        const int amount = sz.height != 1 ? sz.height : sz.width;
```
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/perf/common/gapi_core_perf_tests_inl.hpp]: You are right! Thanks for your attentiveness

File: modules/gapi/test/common/gapi_core_tests_common.hpp (hunk 1 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_core_tests_common.hpp]: Yes, thank you! applied

File: modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp (hunk 4 above)
Comment (OrestChura) [reply modules/gapi/perf/cpu/gapi_core_perf_tests_cpu.cpp]: Forgot to delete, thanks - applied
In fact, the parameters has been changed a bit, please take a look

File: modules/gapi/perf/common/gapi_core_perf_tests_inl.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/perf/common/gapi_core_perf_tests_inl.hpp:1914]: Ok

File: modules/gapi/test/common/gapi_core_tests_common.hpp (hunk 2 above)
Comment (mpashchenkov) [reply modules/gapi/test/common/gapi_core_tests_common.hpp:164]: My bad. Do this tests contain functions which are the same `initmatrixrandu` (for example)?

Comment (OrestChura) [reply modules/gapi/test/common/gapi_core_tests_common.hpp:164]: Yes, so there is an input `cv::Mat` or `std::vector` which I fill in by these our standard functions; but `bestLabels` is an optional input (depending on `KMEANS_USE_INITIAL_LABELS` whether to be needed) so I can't use the standard init function for it and fill it by hand

File: modules/gapi/test/common/gapi_core_tests_common.hpp [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+template<typename Pt>
+void kmeansTestValidate(const cv::Size& sz, const MatType2&, const int K,
+                        const double compact_gapi, const std::vector<int>& labels_gapi,
+                        const std::vector<Pt>& centers_gapi)
+{
+    const int amount = sz.height;
+    // Validation
+    EXPECT_GE(compact_gapi, 0.);
```
Comment (mpashchenkov) [inline modules/gapi/test/common/gapi_core_tests_common.hpp:99]: Is `compact_gapi` always 0?

File: modules/gapi/test/common/gapi_core_tests_common.hpp [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    const cv::TermCriteria criteria(cv::TermCriteria::MAX_ITER + cv::TermCriteria::EPS, 30, 0);
+    const int attempts = 1;
+    Labels labels_ocv;
+    In centers_ocv;
+    { // step to generalize cv::Mat & std::vector cases of bestLabels' types
+        cv::Mat bestLabelsMat(bestLabels);
+        bestLabelsMat.copyTo(labels_ocv);
+    }
+    // OpenCV code /////////////////////////////////////////////////////////////
+    double compact_ocv = cv::kmeans(in, K, labels_ocv, criteria, attempts, flags, centers_ocv);
```
Comment (mpashchenkov) [inline modules/gapi/test/common/gapi_core_tests_common.hpp:136]: const, but it is a cavil.

File: modules/gapi/test/common/gapi_core_tests_common.hpp (hunk 9 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_core_tests_common.hpp:99]: It should always be Greater or Equal than 0

**Review by mpashchenkov — APPROVED** [review]

File: modules/gapi/test/common/gapi_core_tests_common.hpp (hunk 10 above)
Comment (OrestChura) [reply modules/gapi/test/common/gapi_core_tests_common.hpp:136]: Yeah, I don't think this changes much. Let it be non-const

**Comment by OrestChura** [comment]
@alalek, is this mergeable?

**Comment by alalek** [comment]
@dmatveev @dbudniko Please assign delegated "reviewer" into "assignee" field or review yourself this PR (similar note is about all G-API PRs).
